Improve Network Policy Validation

renyunkang opened the pull request:
Description:  https://github.com/cilium/cilium/issues/35558#issuecomment-2513354154

Fixes: https://github.com/cilium/cilium/issues/35558

```release-note
Improves Network Policy validation and default deny behavior. Policies now require at least one of Ingress, IngressDeny, Egress, or EgressDeny to be defined.
```

learnitall commented on pkg/policy/api/rule_validation.go:
```diff
@@ -37,19 +37,32 @@ func (r *Rule) Sanitize() error {
-			x := len(r.Ingress) > 0 || len(r.IngressDeny) > 0
-			r.EnableDefaultDeny.Ingress = &x
+			hasIngress = len(r.Ingress) > 0 || len(r.IngressDeny) > 0
+			if r.EnableDefaultDeny.Egress == nil && !hasEgress && !hasIngress {
```
I think the nil check is redundant since `hasEgress` should be appropriately set if it is nil.
```suggestion
			if !hasEgress && !hasIngress {
```

learnitall reviewed (approved):
The API changes look good to go from my perspective! Thanks!

squeed reviewed (changes requested):
While not perfect, we should also add OneOf constraints to the CRD, so that we reject invalid policies before creation as much as possible.

joestringer commented on pkg/policy/api/rule_validation.go:
```diff
@@ -37,19 +37,32 @@ func (r *Rule) Sanitize() error {
-			r.EnableDefaultDeny.Ingress = &x
+			hasIngress = len(r.Ingress) > 0 || len(r.IngressDeny) > 0
+			if r.EnableDefaultDeny.Egress == nil && !hasEgress && !hasIngress {
+				return fmt.Errorf("rule must have one of Ingress/IngressDeny/Egress/EgressDeny/EnableDefaultDeny.Ingress/EnableDefaultDeny.Egress")
```
I'm inclined to agree with Casey on this: If we declare a `oneOf` requirement in the CRD spec for the ingress/ingressDeny/egress/egressDeny then I don't think we need to change this runtime validation at all, because the misconfiguration would be caught in kube-apiserver.

This does subtly deviate from what you're proposing here, in that it makes the rule simpler: There must always be at least one rule in one of these statements. So rather than "supporting default deny policy with no rules", the proposal would "require at least one rule in every network policy".

joestringer replied:
Hmm, with the [latest discussion point on the CFP issue](https://github.com/cilium/cilium/issues/35558#issuecomment-2503006466), @renyunkang points out a more complicated case that would require something like this runtime check. Not sure if there is a way to solve that problem purely with `kubebuilder` annotations.

joestringer commented on pkg/policy/api/rule_validation_test.go:
```diff
@@ -1174,69 +1178,90 @@ func BenchmarkCIDRSanitize(b *testing.B) {
-		assert.NotNil(t, b.EnableDefaultDeny.Ingress)
-
-		assert.Equal(t, tc.wantEgress, *b.EnableDefaultDeny.Egress, "Rule.EnableDefaultDeny.Egress should match")
-		assert.Equal(t, tc.wantIngress, *b.EnableDefaultDeny.Ingress, "Rule.EnableDefaultDeny.Ingress should match")
```
It looks like this test takes the public version of a rule without `EnableDefaultDeny` and then sanitizes the rule for internal use, which populates the `EnableDefaultDeny.Ingress` if there is an ingress rule (and equivalent for egress).

However now with the latest changes, the "default deny" property is no longer tested but it's being replaced with a test for the number of rules.

I think that ideally this test validates both things - ensure that the default deny behavior is correct post-sanitization even if the user did not explicitly specify it, _and_ validate the error checking for the new cases you've introduced. Please restore the `tc.wantEgress` / `tc.wantIngress` test cases.

renyunkang replied:
done.

joestringer commented on pkg/policy/api/rule_validation.go:
```diff
@@ -33,24 +33,31 @@ var (
-		r.EnableDefaultDeny.Ingress = &enableDefaultDenyDefault
+		if len(r.Egress) > 0 || len(r.EgressDeny) > 0 {
+			r.EnableDefaultDeny.Egress = &enableDefaultDenyDefault
+		}
```
:thinking: I think we need this "if there is an egress rule and the user hasn't specified the default deny behavior, set the default deny" independently of whether `option.Config.EnableNonDefaultDenyPolicies` is enabled. Same for Ingress.

Otherwise after sanitization the `r.EnableDefaultDeny.Ingress` or `r.EnableDefaultDeny.Egress` could still be nil, which may lead to problems later on while handling the field.

I think that the change here to make the "else" (!config) case also check the length of the egress/egressdeny slices is correct and probably a bug in existing releases if the config flag is disabled, but I guess the config flag is not disabled in tests so it's probably poorly tested.

renyunkang replied:
The changes have been rolled back.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -753,7 +753,7 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
 			if len(r.Ingress) > 0 || len(r.IngressDeny) > 0 {
 				ingress = true
-				if *r.EnableDefaultDeny.Ingress {
+				if r.EnableDefaultDeny.Ingress == nil || *r.EnableDefaultDeny.Ingress {
```
I'd suggest sanitizing the rules properly instead, then we shouldn't need extra nil checks much later in processing like this.
```suggestion
				if *r.EnableDefaultDeny.Ingress {
```

renyunkang replied:
done.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -762,7 +762,7 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
 			if len(r.Egress) > 0 || len(r.EgressDeny) > 0 {
 				egress = true
-				if *r.EnableDefaultDeny.Egress {
+				if r.EnableDefaultDeny.Egress == nil || *r.EnableDefaultDeny.Egress {
```
I'd suggest sanitizing the rules properly instead, then we shouldn't need extra nil checks much later in processing like this.
```suggestion
				if *r.EnableDefaultDeny.Egress {
```

renyunkang replied:
done.

joestringer commented on pkg/policy/directory/watcher_test.go:
```diff
@@ -117,15 +117,15 @@ func TestDeleteFromPolicyEngine(t *testing.T) {
 	data := []byte(policy1)
 	cnp, _ := p.translateToCNPObject(data)
-	err := p.addToPolicyEngine(cnp, "test.yaml")
+	err := p.addToPolicyEngine(cnp.DeepCopy(), "test.yaml")
```
I didn't notice the `DeepCopy()` before. Why is this change needed?

joestringer replied:
Ah, I guess this is also a side-effect of the sanitize call not being idempotent. Let's discuss below, but if it's possible to avoid this then it would be nice to not have to make copies in all these other places. https://github.com/cilium/cilium/pull/35904#discussion_r1878897367

joestringer commented on pkg/policy/repository_deny_test.go:
```diff
@@ -58,7 +58,6 @@ func TestComputePolicyDenyEnforcementAndRules(t *testing.T) {
 			fooIngressDenyRule1Label,
 		},
 	}
-	fooIngressDenyRule1.Sanitize()
```
I see the problem you've identified about non-idempotent `Sanitize`. Out of curiousity, what happens if you initialize the `egress` or `ingress` slice to empty iff the sanitize call populates the default deny field? Does that cause other problems in enforcement calculation or would it retain the current network policy behavior? If it retains the behavior, then this might help to avoid needing to modify all these sanitize calls.

renyunkang replied:
> Does that cause other problems in enforcement calculation or would it retain the current network policy behavior? If it retains the behavior, then this might help to avoid needing to modify all these sanitize calls.

Does "initialize the egress or ingress slice to empty" mean creating a non-nil slice with `len(ingress) == 0`, or a slice with one empty element `(len(ingress) == 1)`? I believe the former is correct, which would require an additional nil check. I need to investigate further for potential impacts.

renyunkang replied:
The `egress` or `ingress` slice is initialized to empty if the sanitize call populates the default deny field. This retains the current network policy behavior. The updated code is available here: https://gist.github.com/renyunkang/68b80b4734c5088cec7bc3ca855d87c1

Please confirm whether this modification is acceptable. @joestringer

joestringer replied:
That's what I was wondering about yeah. I think to understand the impact though we would have to run the full testsuite to make sure that this doesn't introduce unexpected policy handling behavior. If you push the proposed branch here then I can trigger the testsuite.

renyunkang committed:
policy: Reject policies if they have no rules statement.

Previously it was ambiguous whether the 'enableDefaultDeny' field in the network policy could be specified without rules. In order to clarify the usage of this field, ensure that at least one of the ingress, ingressDeny, egress, or egressDeny fields are specified by the user in each policy.

Signed-off-by: renyunkang <rykren1998@gmail.com>

joestringer reviewed (approved):
Looks good, thanks for all your work on this @renyunkang.
